src/garbage_collector.rs

```diff
@@ -0,0 +1,253 @@
+            .core
+            .l0
+            .iter()
+            .map(|sst| sst.id)
+            .collect::<HashSet<_>>();
+        let active_compacted_ssts = manifest
+            .core
+            .compacted
+            .iter()
+            .flat_map(|sr| sr.ssts.iter().map(|sst| sst.id))
+            .collect::<HashSet<SsTableId>>();
+        let active_ssts = active_l0_ssts
+            .union(&active_compacted_ssts)
+            .collect::<HashSet<_>>();
+        let min_age = self
+            .options
+            .compacted_options
+            .map_or(DEFAULT_MIN_AGE, |opts| opts.min_age);
+        let sst_ids_to_delete = self
+            .table_store
+            // List all SSTs in the table store
+            .list_compacted_ssts(..)
+            .await?
+            .into_iter()
+            // Filter out the ones that are too young to be collected
+            .filter(|sst| {
+                let min_age = chrono::Duration::from_std(min_age).expect("invalid duration");
+                Utc::now().signed_duration_since(sst.last_modified) > min_age
+            })
+            .map(|sst| sst.id)
+            // Filter out the ones that are active in the manifest
+            .filter(|id| !active_ssts.contains(id))
+            .collect::<Vec<_>>();
+
+        for id in sst_ids_to_delete {
+            if let Err(e) = self.table_store.delete_sst(&id).await {
+                error!("Error deleting SST: {}", e);
+            } else {
+                self.db_stats.gc_compacted_count.inc();
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Run the garbage collector
+    pub async fn run(&self) {
+        let manifest_ticker = Self::options_to_ticker(self.options.manifest_options.as_ref());
+        let wal_ticker = Self::options_to_ticker(self.options.wal_options.as_ref());
+        let compacted_ticker = Self::options_to_ticker(self.options.compacted_options.as_ref());
+
+        loop {
+            crossbeam_channel::select! {
+                recv(manifest_ticker) -> _ => {
+                    if let Err(e) = self.collect_garbage_manifests().await {
+                        error!("Error collecting manifest garbage: {}", e);
+                    }
+                }
+                recv(wal_ticker) -> _ => {
+                    if let Err(e) = self.collect_garbage_wal_ssts().await {
+                        error!("Error collecting WAL garbage: {}", e);
+                    }
+                }
+                recv(compacted_ticker) -> _ => {
+                    if let Err(e) = self.collect_garbage_compacted_ssts().await {
+                        error!("Error collecting compacted garbage: {}", e);
+                    }
+                }
+                recv(self.external_rx) -> _ => break, // Shutdown
+            }
+        }
+    }
+
+    fn options_to_ticker(
+        options: Option<&GarbageCollecterDirectoryOptions>,
+    ) -> crossbeam_channel::Receiver<std::time::Instant> {
```

Comment:
Should we add an (perhaps test only) option that collects garbage on demand, and use that for testing the deletion? `min_age` could be 0 and it would cause all the inactive files to be deleted.

Reply:
I ended up being able to write the tests that just set the options to 0ms for tickers and 3600 fort min age. I then set the expired objects to be older than 3600 using the local filesystem mtime metadata.